### PR TITLE
calicoctl ipam release/show cmd

### PR DESCRIPTION
**Pull request by gunjan5**
Fix for #87

### CODE REVIEW

**Comment by gunjan5** [comment]
Behavior changes from the python version to go version: 

`calicoctl ipam info <IP>` —> `calicoctl ipam show <IP>` 

Python calicoctl:
```bash
vagrant@calico-01:~$ calicoctl ipam info 192.168.142.1
No attributes defined for 192.168.142.1
vagrant@calico-01:~$ calicoctl ipam info 192.168.142.4
IP 192.168.142.4 is not currently assigned
```
```bash 
vagrant@calico-01:~$ calicoctl ipam release 192.168.142.4
Failed to release address
vagrant@calico-01:~$ calicoctl ipam release 192.168.142.2
Address successfully released
```

Go calicoctl:
```bash 
vagrant@calico-01:~$ calicoctl ipam show 192.168.142.1
No attributes defined for IP 192.168.142.1
vagrant@calico-01:~$ calicoctl ipam show 192.168.142.4
IP 192.168.142.4 is not currently assigned in block
```
```bash 
vagrant@calico-01:~$ calicoctl ipam release 192.168.142.4
Failed to release IP address 192.168.142.4
vagrant@calico-01:~$ calicoctl ipam release 192.168.142.3
Successfully released IP address 192.168.142.3
```

File: calicoctl/commands/ipam.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	caliconet "github.com/tigera/libcalico-go/lib/net"
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: lower case `<ip>`?

File: calicoctl/commands/ipam.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    calicoctl ipam show <IP>
+
+Description:
+    Manage Calico assigned IP addresses
+
+Warnings:
+  -  Releasing an in-use IP address can result in it being assigned to multiple
+     workloads.
+`
+
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: kill a few of these linespaces

File: calicoctl/commands/ipam.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		os.Exit(1)
+	}
+
+	return nil
+}
+
+// showIP gets the attributes of an IP address, and returns nil if it is assigned
+// or an error with a message if not assigned
+func showIP(ipamClient client.IPAMInterface, ip net.IP) error {
+
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: extra newline

File: calicoctl/commands/ipam.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	attr, err := ipamClient.GetAssignmentAttributes(caliconet.IP{ip})
+
+	// IP address is assigned
+	if err == nil {
+
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: extra newline

File: calicoctl/commands/ipam.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	return nil
+}
+
+// releaseIP releases the IP address passed to it
+// or prints error message if it's already released
+func releaseIP(ipamClient client.IPAMInterface, ip net.IP) error {
+	ips := []caliconet.IP{}
+	ips = append(ips, caliconet.IP{ip})
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: Can this be one-lined to initialize `ips` with `ip` in it?

File: calicoctl/commands/ipam.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+	// Call ReleaseIPs releases the IP and returns an empty slice as ipAfterRelease if
+	// release was successful else it returns back the slice with the IP passed in
+	ipAfterRelease, err := ipamClient.ReleaseIPs(ips)
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: how about `ipAfterRelease` :arrow_right:  `unreleasedIPs`

File: calicoctl/commands/ipam.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	// if ipAfterRelease slice is empty then IP was released Successfully
+	if len(ipAfterRelease) == 0 {
+		fmt.Printf("Successfully released IP address %s\n", ip)
+	} else {
+
+		// Couldn't release the IP if the slice is not empty
+		// or IP might already be released/unassigned
+		fmt.Printf("Failed to release IP address %s\n", ip)
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: Why not return this as an error so that the caller can print it?

File: calicoctl/commands/ipam.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+		if len(attr) != 0 {
+			fmt.Println(attr)
+		} else {
+			// IP address is assigned but attributes are not set
+			fmt.Printf("No attributes defined for IP %s\n", ip)
+		}
+	} else {
+		// IP address is not assigned, this prints message like
+		// `IP 192.168.71.1 is not assigned in block`
+		fmt.Println(err)
```
Comment (djosborne) [inline calicoctl/commands/ipam.go]: Same here - return `err` for the caller to print?

Comment (gunjan5) [inline calicoctl/commands/ipam.go]: I've restructured the whole conditional logic, but the error can't be returned to the caller because it is not exactly an error (it's more of a message to print. I just added a comment on top to explain it), and in Go, idiomatic way to handle error is to handle them where they happen instead of bubbling them up (we don't have that in our code base everywhere but I'd like to follow that best practice for the new code that gets written)

File: calicoctl/commands/ipam.go (hunk 6 above)
Comment (gunjan5) [inline calicoctl/commands/ipam.go]: sounds good! Spent a lot of time thinking about the variable name but that's the best I could come up with :)

File: calicoctl/commands/ipam.go (hunk 7 above)
Comment (gunjan5) [inline calicoctl/commands/ipam.go]: same as before - in Go, idiomatic way is to handle errors where they happen instead of bubbling them up

**Comment by gunjan5** [comment]
pushed the changes PTAL

File: calicoctl/calicoctl.go [hunk 9]
```diff
@@ -48,7 +49,7 @@ Options:
 	if logLevel := arguments["--log-level"]; logLevel != nil {
 		parsedLogLevel, err := log.ParseLevel(logLevel.(string))
 		if err != nil {
-			fmt.Printf("Unknown log level: %s, expected one of: \n" +
+			fmt.Printf("Unknown log level: %s, expected one of: \n"+
```
Comment (caseydavenport) [inline calicoctl/calicoctl.go]: Could you please undo this change?

File: calicoctl/commands/ipam.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	"net"
+	"os"
+
+	"github.com/tigera/libcalico-go/lib/client"
+	caliconet "github.com/tigera/libcalico-go/lib/net"
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: This comment is wrong - it shouldn't talk about "show/release" (or any other specific IPAM subcommands), since this is the general comment for the entire IPAM func.

File: calicoctl/commands/ipam.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
+    calicoctl ipam show <IP>
+
+Description:
+    Manage Calico assigned IP addresses
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: I'd call this "IP address management".  No need to call this out as "Calico" (since it is part of `calicoctl`)

File: calicoctl/commands/ipam.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  -  Releasing an in-use IP address can result in it being assigned to multiple
+     workloads.
+`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// length of parsedArgs is 0 when `-h` `--help` is passed
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: Make sure you're using capitalization / punctuation in your comments :)

File: calicoctl/commands/ipam.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+	// docopts takes care of printing the help message, but we need to
+	// stop return nil here so command doesn't go any further when `-h` is passed
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	// Parse the input string as an IP address (IPv4 and IPv6)
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: There are only two commands at the moment, each of which expects the second arg to be the IP address.  However, this may not always be the case.  I think argument parsing like this should be the responsibility of the subcommand, yeah?

File: calicoctl/commands/ipam.go [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
+    calicoctl ipam show <IP>
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: `calicoctl ipam show <IP>` is probably better as `calicoctl ipam show --ip=<IP>`.

This means we can be more flexible in terms of the commands we add here later. e.g
```
calicoctl ipam show --node=<node>
```
or something similar.

File: calicoctl/commands/ipam.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+	caliconet "github.com/tigera/libcalico-go/lib/net"
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: Same comment probably applies to this - a `--ip` option is more flexible here.

File: calicoctl/commands/ipam.go [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+	} else {
+		// IP address is assigned but attributes are not set
+		fmt.Printf("No attributes defined for IP %s\n", ip)
+	}
+
+	return nil
+}
+
+// releaseIP releases the IP address passed to it
+// or prints error message if it's already released
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: or prints _an_ error message.

Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: s/already released/not assigned

File: calicoctl/commands/ipam.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+func releaseIP(ipamClient client.IPAMInterface, ip net.IP) error {
+	ips := []caliconet.IP{caliconet.IP{ip}}
+
+	// Call ReleaseIPs releases the IP and returns an empty slice as unreleasedIPs if
+	// release was successful else it returns back the slice with the IP passed in
+	unreleasedIPs, err := ipamClient.ReleaseIPs(ips)
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: I know @djosborne suggested `unreleasedIPs`, but these are actually `unallocatedIPs`.

The IP addresses returned here indicate that they were never allocated in the first place.

File: calicoctl/commands/ipam.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		fmt.Printf("Error: %v\n", err)
+		return err
+	}
+
+	// Couldn't release the IP if the slice is not empty
+	// or IP might already be released/unassigned. This is not exactly an error,
+	// so not returning it to the caller
+	if len(unreleasedIPs) != 0 {
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: This case means that the given IP address wasn't assigned.  The message we print out should indicate that more clearly.

How about "IP address %s is not assigned" ?

Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: If there is actually an error releasing the IP address, that will be handled by the `err != nil` check above.

File: lib/client/ipam_block.go [hunk 19]
```diff
@@ -314,13 +314,13 @@ func (b allocationBlock) attributesForIP(ip cnet.IP) (map[string]string, error)
 	// Convert to an ordinal.
 	ordinal := ipToOrdinal(ip, b)
 	if (ordinal < 0) || (ordinal > blockSize) {
-		return nil, errors.New("IP address not in block")
+		return nil, errors.New(fmt.Sprintf("IP %s not in block", ip))
```
Comment (caseydavenport) [inline lib/client/ipam_block.go]: This log should print out the block as well.

**Comment by caseydavenport** [comment]
A few more comments, then LGTM.

File: calicoctl/calicoctl.go (hunk 9 above)
Comment (gunjan5) [inline calicoctl/calicoctl.go]: not sure I understand what you mean. The spacing was adjusted by `gofmt` (which should have been to begin with)

Comment (caseydavenport) [inline calicoctl/calicoctl.go]: Interesting... I didn't realize that was standard Go formatting.

Looks pretty weird to me, but so be it :(

File: calicoctl/commands/ipam.go (hunk 13 above)
Comment (gunjan5) [inline calicoctl/commands/ipam.go]: ok, created a `validateIP` func and calling it from subcommands now

File: calicoctl/commands/ipam.go (hunk 14 above)
Comment (gunjan5) [inline calicoctl/commands/ipam.go]: > calicoctl ipam show <IP> is probably better 
I assume you meant `calicoctl ipam show --ip=<IP>` :)

**Comment by gunjan5** [comment]
@caseydavenport made those changes and CircleCI build passes now. PTAL

File: calicoctl/commands/ipam.go [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release --ip=<IP>
+    calicoctl ipam show --ip=<IP>
+
+Description:
+    IP address management
+
+Options:
+    --ip=<IP>     The directory for logs [default: /var/log/calico]
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: The directory for logs?  ;)

File: calicoctl/commands/ipam.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// Length of parsedArgs is 0 when `-h` `--help` is passed
+	// docopts takes care of printing the help message, but we need to
+	// stop return nil here so command doesn't go any further when `-h` is passed
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: The punctuation, etc on this comment is still not right.

```
// Length of parsedArgs is 0 when `-h` `--help` is passed.
// Docopt takes care of printing the help message, but we need to
// return nil here so command doesn't go any further when `-h` is passed.
```

File: calicoctl/commands/ipam.go [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	// create a new backend client from env vars
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]:
```
// Create a new backend client from env vars.
```

File: calicoctl/commands/ipam.go [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+	ipamClient := backendClient.IPAM()
+
+	switch args[1] {
+	case "show":
+		err = showIP(ipamClient, parsedArgs["--ip"].(string))
+	case "release":
+		err = releaseIP(ipamClient, parsedArgs["--ip"].(string))
+	}
+
+	// Exit with an error message if any of the command returned an error
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: This guy needs a period at the end.

Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: s/command/commands

File: calicoctl/commands/ipam.go [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+		fmt.Printf("IP address %s is not assigned\n", ip)
+		return nil
+	}
+
+	// If unallocatedIPs slice is empty then IP was released Successfully
+	fmt.Printf("Successfully released IP address %s\n", ip)
+	return nil
+}
+
+func validateIP(str string) net.IP {
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: This could use a GoDoc comment.

File: calicoctl/commands/ipam.go [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+	// Call ReleaseIPs releases the IP and returns an empty slice as unallocatedIPs if
+	// release was successful else it returns back the slice with the IP passed in
+	unallocatedIPs, err := ipamClient.ReleaseIPs(ips)
+	if err != nil {
+		fmt.Printf("Error: %v\n", err)
+		return err
+	}
+
+	// Couldn't release the IP if the slice is not empty
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: Comment needs punctuation.

File: calicoctl/commands/ipam.go [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+// releaseIP releases the IP address passed to it
+// or prints an error message if it's not assigned
+func releaseIP(ipamClient client.IPAMInterface, passedIP string) error {
+	ip := validateIP(passedIP)
+	ips := []caliconet.IP{caliconet.IP{ip}}
+
+	// Call ReleaseIPs releases the IP and returns an empty slice as unallocatedIPs if
+	// release was successful else it returns back the slice with the IP passed in
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: Needs punctuation.

File: calicoctl/commands/ipam.go [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+
+	// Couldn't release the IP if the slice is not empty
+	// or IP might already be released/unassigned. This is not exactly an error,
+	// so not returning it to the caller
+	if len(unallocatedIPs) != 0 {
+		fmt.Printf("IP address %s is not assigned\n", ip)
+		return nil
+	}
+
+	// If unallocatedIPs slice is empty then IP was released Successfully
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: Needs punctuation. :)

**Comment by caseydavenport** [comment]
Some more review comments from me.

I'm enjoying the amount of comments you've added to the code, but make sure you're using proper capitalization and punctuation!

File: calicoctl/commands/ipam.go (hunk 22 above)
Comment (gunjan5) [inline calicoctl/commands/ipam.go]: @caseydavenport not sure if I agree. Technically, periods should only be added to a complete sentence, this comment, like many others, is not a complete sentence.
Ref: http://www.mit.edu/course/21/21.guide/period.htm 

I'm okay with it if we do it for consistency. 

Thoughts?

Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: I believe that link says to put periods at the end of both declarative, and imperative sentences, and this would be an instance of an imperative sentence.  See: http://www.mit.edu/course/21/21.guide/imperati.htm

```
Use periods to indicate the end of declarative or imperative sentences.

The Newcomen engine utilizes the condensation of steam to create a partial vacuum beneath a piston.
Expand the system's bandwidth.
```

:)

**Comment by fasaxc** [comment]
TODO:
- [ ] Fix up imports to use projectcallico instead of tigera

**Comment by caseydavenport** [comment]
@gunjan5 could you address @fasaxc's comment to switch over the imports to use `projectcalico`, just for the files touched in this PR?

Once that is done, LGTM!

**Comment by gunjan5** [comment]
@caseydavenport done! but the build is failing because of that which, I think, is expected. 
Also, commits are squashed

**Comment by caseydavenport** [comment]
> fatal: unable to access 'https://go.googlesource.com/tools/': Failed to connect to go.googlesource.com port 443: Connection timed out

I'm not sure the build is failing because of the import changes.

**Comment by gunjan5** [comment]
@caseydavenport seems like it was a temporary glitch, it's gone now. The build failure is because of different import paths now

**Comment by caseydavenport** [comment]
@gunjan5 I think a rebase against master will fix that for you.

**Comment by gunjan5** [comment]
@caseydavenport done! PTAL

File: calicoctl/commands/ipam.go [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+	if len(unallocatedIPs) != 0 {
+		fmt.Printf("IP address %s is not assigned\n", ip)
+		return
+	}
+
+	// If unallocatedIPs slice is empty then IP was released Successfully.
+	fmt.Printf("Successfully released IP address %s\n", ip)
+}
+
+// validateIP takes a string as an inoput and makes sure it's a valid IPv4 or IPv6 address.
```
Comment (caseydavenport) [inline calicoctl/commands/ipam.go]: typo: inoput
